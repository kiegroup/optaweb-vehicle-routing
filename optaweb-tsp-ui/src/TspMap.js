import PropTypes from 'prop-types';
import React from 'react';
import { Map, Marker, Polygon, Popup, TileLayer, Tooltip, ZoomControl } from 'react-leaflet';

function TspMap({ center, zoom, selectedId, route, domicileId, clickHandler, removeHandler }) {
  return (
    <Map
      center={center}
      zoom={zoom}
      onClick={clickHandler}
      style={{ width: '100vw', height: '100vh' }}
      zoomControl={false} // hide the default zoom control which is on top left
    >
      <TileLayer
        attribution="&amp;copy <a href=&quot;http://osm.org/copyright&quot;>OpenStreetMap</a> contributors"
        url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
      />
      <ZoomControl position="topright" />
      {
        route.map(location => (
          <Marker
            key={location.id}
            position={location}
          >
            <Popup>
              <button
                disabled={route.length > 1 && location.id === domicileId}
                onClick={() => removeHandler(location.id)}
              >x
              </button>
            </Popup>
            <Tooltip
              // The permanent and non-permanent tooltips are different components
              // and need to have different keys
              key={location.id + (location.id === selectedId ? 'T' : 't')}
              permanent={location.id === selectedId}
            >
              {`Location ${location.id} [Lat=${location.lat}, Lng=${location.lng}]`}
            </Tooltip>
          </Marker>
        ))
      }
      <Polygon
        positions={route}
        fill={false}
      />
    </Map>
  );
}

TspMap.propTypes = {
  center: PropTypes.shape({
    lat: PropTypes.number.isRequired,
    lng: PropTypes.number.isRequired,
  }).isRequired,
  zoom: PropTypes.number.isRequired,
  selectedId: PropTypes.number.isRequired,
  route: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    lat: PropTypes.number.isRequired,
    lng: PropTypes.number.isRequired,
  })).isRequired,
  domicileId: PropTypes.number.isRequired,
  clickHandler: PropTypes.func.isRequired,
  removeHandler: PropTypes.func.isRequired,
};

export default TspMap;
