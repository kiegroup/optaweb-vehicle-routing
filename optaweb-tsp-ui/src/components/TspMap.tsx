/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as L from 'leaflet';
import * as React from 'react';
import {
  Map,
  Marker,
  Polyline,
  TileLayer,
  Tooltip,
  ZoomControl,
} from 'react-leaflet';
import { ILatLng, ITSPRouteWithSegments } from '../store/tsp';

interface ITspMapProps extends ITSPRouteWithSegments {
  center: ILatLng;
  zoom: number;
  selectedId: number;
  clickHandler: (e: React.SyntheticEvent<HTMLElement>) => void;
  removeHandler: (id: number) => void;
}

const TspMap: React.SFC<ITspMapProps> = ({
  center,
  zoom,
  selectedId,
  route,
  segments,
  domicileId,
  clickHandler,
  removeHandler,
}) => {
  const homeIcon = L.icon({
    iconAnchor: [12, 12],
    iconSize: [24, 24],
    iconUrl: 'if_big_house-home_2222740.png',
    popupAnchor: [0, -10],
    shadowAnchor: [16, 2],
    shadowSize: [50, 16],
    shadowUrl: 'if_big_house-home_2222740_shadow.png',
  });
  const defaultIcon = new L.Icon.Default();
  return (
    <Map
      center={center}
      zoom={zoom}
      onClick={clickHandler}
      style={{ width: '100vw', height: '100vh' }}
      zoomControl={false} // hide the default zoom control which is on top left
    >
      <TileLayer
        attribution={
          '&amp;copy <a href=\"http://osm.org/copyright\">OpenStreetMap</a> contributors'
        }
        url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
      />
      <ZoomControl position="topright" />
      {route.map(location => (
        <Marker
          key={location.id}
          position={location}
          icon={location.id === domicileId ? homeIcon : defaultIcon}
          onClick={() => removeHandler(location.id)}
        >
          <Tooltip
            // The permanent and non-permanent tooltips are different components
            // and need to have different keys
            key={location.id + (location.id === selectedId ? 'T' : 't')}
            permanent={location.id === selectedId}
          >
            {`Location ${location.id} [Lat=${location.lat}, Lng=${location.lng}]`}
          </Tooltip>
        </Marker>
      ))}
      <Polyline positions={segments} fill={false} />
    </Map>
  );
};

export default TspMap;
