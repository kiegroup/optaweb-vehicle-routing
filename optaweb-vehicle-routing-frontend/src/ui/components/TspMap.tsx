/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as L from 'leaflet';
import * as React from 'react';
import { Map, Marker, Polyline, Rectangle, TileLayer, Tooltip, ZoomControl } from 'react-leaflet';
import { LatLng, Location, RouteWithTrack } from 'store/route/types';

type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export interface TspMapProps {
  selectedId: number;
  clickHandler: (e: React.SyntheticEvent<HTMLElement>) => void;
  removeHandler: (id: number) => void;
  depot: Location | null;
  routes: Omit<RouteWithTrack, 'vehicle'>[];
  boundingBox: [LatLng, LatLng] | null;
}

// TODO unlimited unique (random) colors
const colors = ['deepskyblue', 'crimson', 'seagreen', 'slateblue', 'gold', 'darkorange'];

function color(index: number) {
  return colors[index % colors.length];
}

const homeIcon = L.icon({
  iconAnchor: [12, 12],
  iconSize: [24, 24],
  iconUrl: 'if_big_house-home_2222740.png',
  popupAnchor: [0, -10],
  shadowAnchor: [16, 2],
  shadowSize: [50, 16],
  shadowUrl: 'if_big_house-home_2222740_shadow.png',
});

const defaultIcon = new L.Icon.Default();

const marker = (
  removeHandler: (id: number) => void,
  selectedId: number,
  location: Location,
  isDepot: boolean,
) => {
  const icon = isDepot ? homeIcon : defaultIcon;
  return (
    <Marker
      key={location.id}
      position={location}
      icon={icon}
      onClick={() => removeHandler(location.id)}
    >
      <Tooltip
        // The permanent and non-permanent tooltips are different components
        // and need to have different keys
        key={location.id + (location.id === selectedId ? 'T' : 't')}
        permanent={location.id === selectedId}
      >
        {`Location ${location.id} [Lat=${location.lat}, Lng=${location.lng}]`}
      </Tooltip>
    </Marker>
  );
};

const TspMap: React.FC<TspMapProps> = ({
  boundingBox,
  selectedId,
  depot,
  routes,
  clickHandler,
  removeHandler,
}) => {
  const bounds = boundingBox ? new L.LatLngBounds(boundingBox[0], boundingBox[1]) : undefined;
  const center = bounds ? undefined : new L.LatLng(0, 0);
  const zoom = bounds ? undefined : 2;
  return (
    <Map
      bounds={bounds}
      center={center}
      zoom={zoom}
      onClick={clickHandler}
      // FIXME use height: 100%
      style={{ width: '100%', height: 'calc(100vh - 176px)' }}
      zoomControl={false} // hide the default zoom control which is on top left
    >
      <TileLayer
        attribution="&copy; <a href='http://osm.org/copyright'>OpenStreetMap</a> contributors"
        url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
      />
      <ZoomControl position="topright" />
      {depot && marker(removeHandler, selectedId, depot, true)}
      {routes.map(route => (
        route.visits.map(location => marker(removeHandler, selectedId, location, false))
      ))}
      {routes.map((route, index) => (
        <Polyline
          // eslint-disable-next-line react/no-array-index-key
          key={index} // FIXME use unique id (not iteration index)
          positions={route.track}
          fill={false}
          color={color(index)}
        />
      ))}
      {bounds && (
        <Rectangle
          bounds={bounds}
          color="seagreen"
          fill={false}
          dashArray="10,5"
          weight={1}
        />
      )}
    </Map>
  );
};

export default TspMap;
