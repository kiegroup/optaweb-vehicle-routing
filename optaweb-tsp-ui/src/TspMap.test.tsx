/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import React from 'react';
import TspMap from './TspMap';

describe('TSP Map View', () => {
  it('should render correctly', () => {
    const props = {
      center: {
        lat: 1.345678,
        lng: 1.345678,
      },
      route: [
        {
          id: 1,
          lat: 1.345678,
          lng: 1.345678,
        },
        {
          id: 2,
          lat: 2.345678,
          lng: 2.345678,
        },
        {
          id: 3,
          lat: 3.676111,
          lng: 3.568333,
        },
      ],
      domicileId: 1,
      distance: '10',
      clickHandler: jest.fn(),
      removeHandler: jest.fn(),
      zoom: 5,
      selectedId: 1,
    };
    expect.assertions(1);
    const tspMap = shallow(<TspMap {...props} />);
    expect(tspMap).toMatchSnapshot();
  });
});
