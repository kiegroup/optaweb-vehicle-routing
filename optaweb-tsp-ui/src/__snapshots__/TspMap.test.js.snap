// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TSP Map View should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TspMap
    center={
        Object {
            "lat": 1.345678,
            "lng": 1.345678,
          }
    }
    clickHandler={[Function]}
    distance="10"
    domicileId={1}
    removeHandler={[Function]}
    route={
        Array [
            Object {
              "id": 1,
              "lat": 1.345678,
              "lng": 1.345678,
            },
            Object {
              "id": 2,
              "lat": 2.345678,
              "lng": 2.345678,
            },
            Object {
              "id": 3,
              "lat": 3.676111,
              "lng": 3.568333,
            },
          ]
    }
    selectedId={1}
    zoom={5}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "center": Object {
        "lat": 1.345678,
        "lng": 1.345678,
      },
      "children": Array [
        <Unknown
          attribution="&copy <a href=\\"http://osm.org/copyright\\">OpenStreetMap</a> contributors"
          url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
/>,
        <Unknown
          position="topright"
/>,
        Array [
          <Unknown
            icon={
                        NewClass {
                                    "_initHooksCalled": true,
                                    "options": Object {
                                      "iconAnchor": Array [
                                        12,
                                        12,
                                      ],
                                      "iconSize": Array [
                                        24,
                                        24,
                                      ],
                                      "iconUrl": "if_big_house-home_2222740.png",
                                      "popupAnchor": Array [
                                        0,
                                        -10,
                                      ],
                                      "shadowAnchor": Array [
                                        16,
                                        2,
                                      ],
                                      "shadowSize": Array [
                                        50,
                                        16,
                                      ],
                                      "shadowUrl": "if_big_house-home_2222740_shadow.png",
                                    },
                                  }
            }
            onClick={[Function]}
            position={
                        Object {
                                    "id": 1,
                                    "lat": 1.345678,
                                    "lng": 1.345678,
                                  }
            }
>
            <Unknown
                        permanent={true}
            >
                        Location 1 [Lat=1.345678, Lng=1.345678]
            </Unknown>
</Unknown>,
          <Unknown
            icon={
                        NewClass {
                                    "_initHooksCalled": true,
                                    "options": Object {},
                                  }
            }
            onClick={[Function]}
            position={
                        Object {
                                    "id": 2,
                                    "lat": 2.345678,
                                    "lng": 2.345678,
                                  }
            }
>
            <Unknown
                        permanent={false}
            >
                        Location 2 [Lat=2.345678, Lng=2.345678]
            </Unknown>
</Unknown>,
          <Unknown
            icon={
                        NewClass {
                                    "_initHooksCalled": true,
                                    "options": Object {},
                                  }
            }
            onClick={[Function]}
            position={
                        Object {
                                    "id": 3,
                                    "lat": 3.676111,
                                    "lng": 3.568333,
                                  }
            }
>
            <Unknown
                        permanent={false}
            >
                        Location 3 [Lat=3.676111, Lng=3.568333]
            </Unknown>
</Unknown>,
        ],
        <Unknown
          fill={false}
          positions={
                    Array [
                              Object {
                                "id": 1,
                                "lat": 1.345678,
                                "lng": 1.345678,
                              },
                              Object {
                                "id": 2,
                                "lat": 2.345678,
                                "lng": 2.345678,
                              },
                              Object {
                                "id": 3,
                                "lat": 3.676111,
                                "lng": 3.568333,
                              },
                            ]
          }
/>,
      ],
      "onClick": [Function],
      "style": Object {
        "height": "100vh",
        "width": "100vw",
      },
      "zoom": 5,
      "zoomControl": false,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "attribution": "&copy <a href=\\"http://osm.org/copyright\\">OpenStreetMap</a> contributors",
          "url": "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "position": "topright",
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "children": <Unknown
            permanent={true}
>
            Location 1 [Lat=1.345678, Lng=1.345678]
</Unknown>,
          "icon": NewClass {
            "_initHooksCalled": true,
            "options": Object {
              "iconAnchor": Array [
                12,
                12,
              ],
              "iconSize": Array [
                24,
                24,
              ],
              "iconUrl": "if_big_house-home_2222740.png",
              "popupAnchor": Array [
                0,
                -10,
              ],
              "shadowAnchor": Array [
                16,
                2,
              ],
              "shadowSize": Array [
                50,
                16,
              ],
              "shadowUrl": "if_big_house-home_2222740_shadow.png",
            },
          },
          "onClick": [Function],
          "position": Object {
            "id": 1,
            "lat": 1.345678,
            "lng": 1.345678,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": "1T",
          "nodeType": "function",
          "props": Object {
            "children": "Location 1 [Lat=1.345678, Lng=1.345678]",
            "permanent": true,
          },
          "ref": null,
          "rendered": "Location 1 [Lat=1.345678, Lng=1.345678]",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "children": <Unknown
            permanent={false}
>
            Location 2 [Lat=2.345678, Lng=2.345678]
</Unknown>,
          "icon": NewClass {
            "_initHooksCalled": true,
            "options": Object {},
          },
          "onClick": [Function],
          "position": Object {
            "id": 2,
            "lat": 2.345678,
            "lng": 2.345678,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": "2t",
          "nodeType": "function",
          "props": Object {
            "children": "Location 2 [Lat=2.345678, Lng=2.345678]",
            "permanent": false,
          },
          "ref": null,
          "rendered": "Location 2 [Lat=2.345678, Lng=2.345678]",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "function",
        "props": Object {
          "children": <Unknown
            permanent={false}
>
            Location 3 [Lat=3.676111, Lng=3.568333]
</Unknown>,
          "icon": NewClass {
            "_initHooksCalled": true,
            "options": Object {},
          },
          "onClick": [Function],
          "position": Object {
            "id": 3,
            "lat": 3.676111,
            "lng": 3.568333,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": "3t",
          "nodeType": "function",
          "props": Object {
            "children": "Location 3 [Lat=3.676111, Lng=3.568333]",
            "permanent": false,
          },
          "ref": null,
          "rendered": "Location 3 [Lat=3.676111, Lng=3.568333]",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fill": false,
          "positions": Array [
            Object {
              "id": 1,
              "lat": 1.345678,
              "lng": 1.345678,
            },
            Object {
              "id": 2,
              "lat": 2.345678,
              "lng": 2.345678,
            },
            Object {
              "id": 3,
              "lat": 3.676111,
              "lng": 3.568333,
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "center": Object {
          "lat": 1.345678,
          "lng": 1.345678,
        },
        "children": Array [
          <Unknown
            attribution="&copy <a href=\\"http://osm.org/copyright\\">OpenStreetMap</a> contributors"
            url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
/>,
          <Unknown
            position="topright"
/>,
          Array [
            <Unknown
              icon={
                            NewClass {
                                          "_initHooksCalled": true,
                                          "options": Object {
                                            "iconAnchor": Array [
                                              12,
                                              12,
                                            ],
                                            "iconSize": Array [
                                              24,
                                              24,
                                            ],
                                            "iconUrl": "if_big_house-home_2222740.png",
                                            "popupAnchor": Array [
                                              0,
                                              -10,
                                            ],
                                            "shadowAnchor": Array [
                                              16,
                                              2,
                                            ],
                                            "shadowSize": Array [
                                              50,
                                              16,
                                            ],
                                            "shadowUrl": "if_big_house-home_2222740_shadow.png",
                                          },
                                        }
              }
              onClick={[Function]}
              position={
                            Object {
                                          "id": 1,
                                          "lat": 1.345678,
                                          "lng": 1.345678,
                                        }
              }
>
              <Unknown
                            permanent={true}
              >
                            Location 1 [Lat=1.345678, Lng=1.345678]
              </Unknown>
</Unknown>,
            <Unknown
              icon={
                            NewClass {
                                          "_initHooksCalled": true,
                                          "options": Object {},
                                        }
              }
              onClick={[Function]}
              position={
                            Object {
                                          "id": 2,
                                          "lat": 2.345678,
                                          "lng": 2.345678,
                                        }
              }
>
              <Unknown
                            permanent={false}
              >
                            Location 2 [Lat=2.345678, Lng=2.345678]
              </Unknown>
</Unknown>,
            <Unknown
              icon={
                            NewClass {
                                          "_initHooksCalled": true,
                                          "options": Object {},
                                        }
              }
              onClick={[Function]}
              position={
                            Object {
                                          "id": 3,
                                          "lat": 3.676111,
                                          "lng": 3.568333,
                                        }
              }
>
              <Unknown
                            permanent={false}
              >
                            Location 3 [Lat=3.676111, Lng=3.568333]
              </Unknown>
</Unknown>,
          ],
          <Unknown
            fill={false}
            positions={
                        Array [
                                    Object {
                                      "id": 1,
                                      "lat": 1.345678,
                                      "lng": 1.345678,
                                    },
                                    Object {
                                      "id": 2,
                                      "lat": 2.345678,
                                      "lng": 2.345678,
                                    },
                                    Object {
                                      "id": 3,
                                      "lat": 3.676111,
                                      "lng": 3.568333,
                                    },
                                  ]
            }
/>,
        ],
        "onClick": [Function],
        "style": Object {
          "height": "100vh",
          "width": "100vw",
        },
        "zoom": 5,
        "zoomControl": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "attribution": "&copy <a href=\\"http://osm.org/copyright\\">OpenStreetMap</a> contributors",
            "url": "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "position": "topright",
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "children": <Unknown
              permanent={true}
>
              Location 1 [Lat=1.345678, Lng=1.345678]
</Unknown>,
            "icon": NewClass {
              "_initHooksCalled": true,
              "options": Object {
                "iconAnchor": Array [
                  12,
                  12,
                ],
                "iconSize": Array [
                  24,
                  24,
                ],
                "iconUrl": "if_big_house-home_2222740.png",
                "popupAnchor": Array [
                  0,
                  -10,
                ],
                "shadowAnchor": Array [
                  16,
                  2,
                ],
                "shadowSize": Array [
                  50,
                  16,
                ],
                "shadowUrl": "if_big_house-home_2222740_shadow.png",
              },
            },
            "onClick": [Function],
            "position": Object {
              "id": 1,
              "lat": 1.345678,
              "lng": 1.345678,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": "1T",
            "nodeType": "function",
            "props": Object {
              "children": "Location 1 [Lat=1.345678, Lng=1.345678]",
              "permanent": true,
            },
            "ref": null,
            "rendered": "Location 1 [Lat=1.345678, Lng=1.345678]",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "children": <Unknown
              permanent={false}
>
              Location 2 [Lat=2.345678, Lng=2.345678]
</Unknown>,
            "icon": NewClass {
              "_initHooksCalled": true,
              "options": Object {},
            },
            "onClick": [Function],
            "position": Object {
              "id": 2,
              "lat": 2.345678,
              "lng": 2.345678,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": "2t",
            "nodeType": "function",
            "props": Object {
              "children": "Location 2 [Lat=2.345678, Lng=2.345678]",
              "permanent": false,
            },
            "ref": null,
            "rendered": "Location 2 [Lat=2.345678, Lng=2.345678]",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "function",
          "props": Object {
            "children": <Unknown
              permanent={false}
>
              Location 3 [Lat=3.676111, Lng=3.568333]
</Unknown>,
            "icon": NewClass {
              "_initHooksCalled": true,
              "options": Object {},
            },
            "onClick": [Function],
            "position": Object {
              "id": 3,
              "lat": 3.676111,
              "lng": 3.568333,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": "3t",
            "nodeType": "function",
            "props": Object {
              "children": "Location 3 [Lat=3.676111, Lng=3.568333]",
              "permanent": false,
            },
            "ref": null,
            "rendered": "Location 3 [Lat=3.676111, Lng=3.568333]",
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "fill": false,
            "positions": Array [
              Object {
                "id": 1,
                "lat": 1.345678,
                "lng": 1.345678,
              },
              Object {
                "id": 2,
                "lat": 2.345678,
                "lng": 2.345678,
              },
              Object {
                "id": 3,
                "lat": 3.676111,
                "lng": 3.568333,
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
